Source/JavaScriptCore/runtime/PropertyTable.h

```diff
@@ -479,59 +512,104 @@ inline PropertyTable* PropertyTable::copy(VM& vm, unsigned newCapacity)
 #ifndef NDEBUG
 inline size_t PropertyTable::sizeInMemory()
 {
-    size_t result = sizeof(PropertyTable) + dataSize();
+    size_t result = sizeof(PropertyTable) + dataSize(isCompact());
     if (m_deletedOffsets)
         result += (m_deletedOffsets->capacity() * sizeof(PropertyOffset));
     return result;
 }
 #endif
 
-inline void PropertyTable::reinsert(const ValueType& entry)
+template<typename Index, typename Entry>
+inline void PropertyTable::reinsert(Index* indexVector, Entry* table, const ValueType& entry)
 {
 #if DUMP_PROPERTYMAP_STATS
     ++propertyTableStats->numReinserts;
 #endif
 
     // Used to insert a value known not to be in the table, and where
     // we know capacity to be available.
-    ASSERT(canInsert());
-    find_iterator iter = find(entry.key);
-    ASSERT(!iter.first);
+    ASSERT(canInsert(entry));
+    auto [emptyIndex, index, offset, attributes] = findImpl(indexVector, table, entry.key());
+    ASSERT_UNUSED(offset, offset == invalidOffset);
+    ASSERT_UNUSED(emptyIndex, emptyIndex == EmptyEntryIndex);
 
     unsigned entryIndex = usedCount() + 1;
-    m_index[iter.second] = entryIndex;
-    table()[entryIndex - 1] = entry;
+    indexVector[index] = entryIndex;
+    table[entryIndex - 1] = entry;
 
     ++m_keyCount;
 }
 
-inline void PropertyTable::rehash(VM& vm, unsigned newCapacity)
+inline void PropertyTable::rehash(VM& vm, unsigned newCapacity, bool forceNonCompact)
```

Comment:
nit: we could invert this bool and call it something like `canStayCompact` so we don't have double negatives in the code below